packages/language/src/preprocessor/pli-preprocessor-parser.ts

```diff
@@ -399,416 +575,583 @@ export class PliPreprocessorParser {
+    const infixOperatorItem: IntermediateBinaryExpression = {
+      infix: true,
+      items: [],
+      operators: [],
+    };
+    infixOperatorItem.items.push(this.primary(state));
     while (true) {
-      let operator: Ops | undefined = undefined;
-      for (const [op, tokenType] of Object.entries<TokenType>(operators)) {
-        if (state.tryConsume(tokenType)) {
-          operator = op as Ops;
+      let operator: string | undefined = undefined;
+      for (const tokenType of PreprocessorBinaryTokens) {
+        if (
+          state.tryConsume(
+            infixOperatorItem as any,
+            CstNodeKind.BinaryExpression_Operator,
+            tokenType,
+          )
+        ) {
+          operator = tokenType.name;
           break;
         }
       }
       if (operator) {
-        const rhs = next.call(this, state);
-        lhs = {
-          type: "binary",
-          lhs,
-          rhs,
-          operator,
-        };
+        const item = this.primary(state);
+        infixOperatorItem.items.push(item);
+        infixOperatorItem.operators.push(operator);
       } else {
         break;
       }
     }
-    return lhs;
+    return constructBinaryExpression(infixOperatorItem);
   }
 
-  private rightAssociativeOperators<Ops extends PPBinaryExpression["operator"]>(
-    state: PreprocessorParserState,
-    operators: Record<Ops, TokenType>,
-    next: (state: PreprocessorParserState) => PPExpression,
-    self: (state: PreprocessorParserState) => PPExpression,
-  ) {
-    let lhs = next.call(this, state);
-    let operator: Ops | undefined = undefined;
-    for (const [op, tokenType] of Object.entries<TokenType>(operators)) {
-      if (state.tryConsume(tokenType)) {
-        operator = op as Ops;
-        break;
-      }
-    }
-    if (operator) {
-      const rhs = self.call(this, state);
-      lhs = {
-        type: "binary",
-        lhs,
-        rhs,
-        operator,
-      };
-    }
-    return lhs;
-  }
-
-  primary(state: PreprocessorParserState): PPExpression {
+  primary(state: PreprocessorParserState): ast.Expression {
     if (state.canConsume(PreprocessorTokens.Number)) {
-      const number = state.consume(PreprocessorTokens.Number);
-      return {
-        type: "number",
-        value: parseInt(number.image, 10), //TODO when to parse binary?
-      };
+      return this.numberLiteral(state);
     } else if (state.canConsume(PreprocessorTokens.String)) {
-      const character = state.consume(PreprocessorTokens.String);
-      const content = this.unpackCharacterValue(character.image);
-      const tokens = this.tokenizePurePliCode(content);
-      return {
-        type: "string",
-        value: tokens,
-      };
+      return this.stringLiteral(state);
     } else if (state.canConsumeKeyword(PreprocessorTokens.Id)) {
       if (state.isOnlyInStatement()) {
-        state.consume(PreprocessorTokens.Percentage);
+        state.consume(
+          undefined!,
```

Comment:
Noticed this showing up quite a bit, Maybe we should relax `consume` to take `SyntaxNode | undefined` up front.